File: src/main/java/com/todoroo/astrid/api/CustomFilter.java
package com.todoroo.astrid.api;

import android.content.ContentValues;
import android.os.Parcel;
import android.os.Parcelable;

import com.todoroo.andlib.utility.AndroidUtilities;
import com.todoroo.astrid.core.SavedFilter;
import com.todoroo.astrid.data.StoreObject;

public class CustomFilter extends Filter {
    private long id;

    private CustomFilter() {

    }

    public CustomFilter(String listingTitle, String sql, ContentValues contentValues, long id) {
        super(listingTitle, sql, contentValues);
        this.id = id;
    }

    public StoreObject toStoreObject() {
        return new StoreObject() {{
            setId(id);
            setValue(SavedFilter.NAME, listingTitle);
            setValue(SavedFilter.SQL, sqlQuery);
            if (valuesForNewTasks != null && valuesForNewTasks.size() > 0) {
                setValue(SavedFilter.VALUES, AndroidUtilities.contentValuesToSerializedString(valuesForNewTasks));
            }
        }};
    }

    public long getId() {
        return id;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void writeToParcel(Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeLong(id);
    }

    @Override
    public void readFromParcel(Parcel source) {
        super.readFromParcel(source);
        id = source.readLong();
    }

    /**
     * Parcelable Creator Object
     */
    public static final Parcelable.Creator<CustomFilter> CREATOR = new Parcelable.Creator<CustomFilter>() {

        /**
         * {@inheritDoc}
         */
        @Override
        public CustomFilter createFromParcel(Parcel source) {
            CustomFilter item = new CustomFilter();
            item.readFromParcel(source);
            return item;
        }

        /**
         * {@inheritDoc}
         */
        @Override
        public CustomFilter[] newArray(int size) {
            return new CustomFilter[size];
        }
    };
}

File: app/src/main/java/com/todoroo/astrid/api/CustomFilter.kt
package com.todoroo.astrid.api

import com.todoroo.astrid.api.Filter.Companion.NO_COUNT
import kotlinx.parcelize.Parcelize
import org.tasks.themes.CustomIcons

@Parcelize
data class CustomFilter(
    val filter: org.tasks.data.Filter,
    override var count: Int = NO_COUNT,
) : Filter {
    override val title: String?
        get() = filter.title
    override val sql: String
        get() = filter.sql!!

    override val valuesForNewTasks: String?
        get() = filter.values

    val criterion: String?
        get() = filter.criterion

    override val order: Int
        get() = filter.order

    val id: Long
        get() = filter.id
    override val icon: Int
        get() = filter.icon ?: CustomIcons.FILTER
    override val tint: Int
        get() = filter.color ?: 0

    override fun areItemsTheSame(other: FilterListItem): Boolean {
        return other is CustomFilter && id == other.id
    }
}
